source/server/listener_manager_impl.h

```diff
@@ -34,15 +36,79 @@ class ProdListenerComponentFactory : public ListenerComponentFactory,
                           Configuration::FactoryContext& context) override {
     return createFilterFactoryList_(filters, server_, context);
   }
-  Network::ListenSocketPtr createListenSocket(Network::Address::InstanceConstSharedPtr address,
-                                              bool bind_to_port) override;
+  Network::ListenSocketSharedPtr
+  createListenSocket(Network::Address::InstanceConstSharedPtr address, bool bind_to_port) override;
   uint64_t nextListenerTag() override { return next_listener_tag_++; }
 
 private:
   Instance& server_;
   uint64_t next_listener_tag_{1};
 };
 
+class ListenerImpl;
+typedef std::unique_ptr<ListenerImpl> ListenerImplPtr;
+
+/**
+ * Implementation of ListenerManager.
+ */
+class ListenerManagerImpl : public ListenerManager, Logger::Loggable<Logger::Id::config> {
+public:
+  ListenerManagerImpl(Instance& server, ListenerComponentFactory& listener_factory,
+                      WorkerFactory& worker_factory);
+
+  void onListenerWarmed(ListenerImpl& listener);
+
+  // Server::ListenerManager
+  bool addOrUpdateListener(const Json::Object& json) override;
+  std::vector<std::reference_wrapper<Listener>> listeners() override;
+  uint64_t numConnections() override;
+  bool removeListener(const std::string& listener_name) override;
+  void startWorkers(GuardDog& guard_dog) override;
+  void stopListeners() override;
+  void stopWorkers() override;
+
+  Instance& server_;
+  ListenerComponentFactory& factory_;
+
+private:
+  typedef std::list<ListenerImplPtr> ListenerList;
+
+  struct DrainingListener {
+    DrainingListener(ListenerImplPtr&& listener, uint64_t workers_pending_removal)
+        : listener_(std::move(listener)), workers_pending_removal_(workers_pending_removal) {}
+
+    ListenerImplPtr listener_;
+    uint64_t workers_pending_removal_;
+  };
+
+  /**
+   * Mark a listener for draining. The listener will no longer be considered active but will remain
+   * present to allow connection draining.
+   * @param listener supplies the listener to drain.
+   */
+  void drainListener(ListenerImplPtr&& listener);
+
+  /**
+   * Get a listener by name. This routine is used because listeners have inherent order in static
+   * configuration and especially for tests. Thus, we can't use a map.
+   * @param listeners supplies the listener list to look in.
+   * @param name supplies the name to search for.
+   */
+  ListenerList::iterator getListenerByName(ListenerList& listeners, const std::string& name);
+
+  ListenerList active_listeners_;
+  ListenerList warming_listeners_;
```

Comment:
Can you document the concept of warming here? Also state the invariants between the {active,warming,draining}_listeners_ lists?